src/player.cpp

```diff
@@ -14798,3 +14804,592 @@ bool player::query_yn( const char *mes, ... ) const
     va_end( ap );
     return ret;
 }
+
+morale_type player::allergy_type( const item &food ) const
+{
+    using allergy_tuple = std::tuple<std::string, std::string, morale_type>;
+    static const std::array<allergy_tuple, 8> allergy_tuples = {{
+        {"VEGETARIAN", "flesh", MORALE_VEGETARIAN},
+        {"VEGETARIAN", "hflesh", MORALE_VEGETARIAN},
+        {"VEGETARIAN", "iflesh", MORALE_VEGETARIAN},
+        {"MEATARIAN", "veggy", MORALE_MEATARIAN},
+        {"LACTOSE", "milk", MORALE_LACTOSE},
+        {"ANTIFRUIT", "fruit", MORALE_ANTIFRUIT},
+        {"ANTIJUNK", "junk", MORALE_ANTIJUNK},
+        {"ANTIWHEAT", "wheat", MORALE_ANTIWHEAT}
+    }};
+
+    for( const auto &tp : allergy_tuples ) {
+        if( has_trait( std::get<0>( tp ) ) &&
+            food.made_of( std::get<1>( tp ) ) ) {
+            return std::get<2>( tp );
+        }
+    }
+
+    return MORALE_NULL;
+}
+
+bool player::is_allergic( const item &food ) const
+{
+    return allergy_type( food ) == MORALE_NULL;
+}
+
+edible_rating player::can_eat( const item &food, bool interactive, bool force ) const
+{
+    if( is_npc() ) {
+        // Just to be sure
+        interactive = false;
+    }
+
+    const char *itname = food.tname().c_str();
```

Comment:
`tname` returns a temporary, `c_str` does as well. Both are per invalid after the full expression. In other words: don't store the result of `c_str`, store the result of `tname` as string (not reference).